nixos/modules/services/misc/pufferpanel.nix

```diff
@@ -0,0 +1,621 @@
+{ config, pkgs, lib, ... }:
+let
+  cfg = config.services.pufferpanel;
+
+  ipwrap = address:
+    if lib.hasInfix ":" address
+    then "[" + address + "]"
+    else address;
+
+  joinHostPort = host: port:
+    "${ipwrap host}:${toString port}";
+
+  toStringOrNull = x:
+    if lib.isBool x then
+      lib.boolToString x
+    else if x != null then
+      toString x
+    else
+      null;
+
+  toJSONOrNull = x:
+    if x != null then
+      builtins.toJSON x
+    else
+      null;
+
+  mkOptionNullable = { type, default ? null, ... }@args:
+    lib.mkOption (args // {
+      type = lib.types.nullOr type;
+      inherit default;
+    });
```

Comment:
These feel *very* out of place: this is one module.

Reply:
I’ve removed options that use `toJSONOrNull` and `mkOptionNullable`, leaving us with `ipwrap` (named after similar Ansible function), `joinHostPort` (named after `net.JoinHostPort` in Go) and `bootToStringOrNull`. Of those, `ipwrap` and `joinHostPort` are good candidates for `lib`.

Reply:
Removed those functions after https://github.com/NixOS/nixpkgs/pull/225274#discussion_r1161951357 discussion.